src/edgehog_ota.c

```diff
@@ -42,331 +42,479 @@
-        nvs_set_u8(handle, OTA_STATE_KEY, OTA_FAILED);
-        nvs_commit(handle);
-        publish_ota_data(astarte_device, req_uuid, OTA_FAILED, EDGEHOG_ERR);
+    uint8_t ota_state;
+    esp_err = nvs_get_u8(handle_nvs, OTA_STATE_KEY, &ota_state);
+    if (esp_err != ESP_OK) {
+        goto end;
+    } else if (ota_state != OTA_STATE_REBOOT) {
+        pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_FAILURE, 0, EDGEHOG_ERR_OTA_INTERNAL);
+        esp_event_post(EDGEHOG_EVENTS, EDGEHOG_OTA_FAILED_EVENT, NULL, 0, 0);
         goto end;
     }
 
-    if (is_partition_changed(handle)) {
-        publish_ota_data(astarte_device, req_uuid, OTA_SUCCESS, EDGEHOG_OK);
+    // Step 4 check if the previous partition is different from the current one.
+    // If not notify astarte of the error.
+
+    uint32_t prev_partition_addr;
+    const esp_partition_t *running_partition_info = esp_ota_get_running_partition();
+    esp_err = nvs_get_u32(handle_nvs, OTA_PARTITION_ADDR_KEY, &prev_partition_addr);
+    if (esp_err == ESP_OK && prev_partition_addr != running_partition_info->address) {
+        pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_SUCCESS, 0, EDGEHOG_OK);
+        esp_event_post(EDGEHOG_EVENTS, EDGEHOG_OTA_SUCCESS_EVENT, NULL, 0, 0);
     } else {
         ESP_LOGE(TAG, "Unable to switch into updated partition");
-        nvs_set_u8(handle, OTA_STATE_KEY, OTA_FAILED);
-        nvs_commit(handle);
-        publish_ota_data(astarte_device, req_uuid, OTA_FAILED, EDGEHOG_ERR_OTA_WRONG_PARTITION);
+        pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_FAILURE, 0, EDGEHOG_ERR_OTA_SYSTEM_ROLLBACK);
+        esp_event_post(EDGEHOG_EVENTS, EDGEHOG_OTA_FAILED_EVENT, NULL, 0, 0);
     }
 
 end:
-    nvs_erase_key(handle, OTA_REQUEST_ID_KEY);
-    nvs_set_u8(handle, OTA_STATE_KEY, OTA_IDLE);
-    nvs_commit(handle);
-    nvs_close(handle);
+    nvs_erase_key(handle_nvs, OTA_REQUEST_ID_KEY);
+    nvs_set_u8(handle_nvs, OTA_STATE_KEY, OTA_STATE_IDLE);
+    nvs_commit(handle_nvs);
+    nvs_close(handle_nvs);
 }
 
-// Beware this function blocks the caller until OTA is completed.
 edgehog_err_t edgehog_ota_event(
-    edgehog_device_handle_t edgehog_device, astarte_device_data_event_t *event_request)
+    edgehog_device_handle_t edgehog_dev, astarte_device_data_event_t *event_request)
 {
-    EDGEHOG_VALIDATE_INCOMING_DATA(TAG, event_request, "/request", BSON_TYPE_DOCUMENT);
-
     uint8_t type;
     uint32_t str_value_len;
 
+    EDGEHOG_VALIDATE_INCOMING_DATA(TAG, event_request, "/request", BSON_TYPE_DOCUMENT);
+
+    // Step 1 get UUID, URL and operation from Astarte request
+
     const void *found = astarte_bson_key_lookup("uuid", event_request->bson_value, &type);
-    const char *request_uuid = astarte_bson_value_to_string(found, &str_value_len);
-    if (!request_uuid || type != BSON_TYPE_STRING) {
+    const char *req_uuid = astarte_bson_value_to_string(found, &str_value_len);
+    if (req_uuid == NULL || type != BSON_TYPE_STRING) {
```

Comment:
Why enforce the NULL check? we prefer the form  `if (ptr) or if (!ptr)` they 
do not expose the opportunity for the accidental assignment. And they are clear and succinct.
```suggestion
    if (req_uuid  || type != BSON_TYPE_STRING) {
```

Reply:
P.S. You meant `if (!req_uuid  || type != BSON_TYPE_STRING) {` right?